File: 백준/Silver/1991. 트리 순회/트리 순회.cc
#include <iostream>
#include <vector>
using namespace std;

int N;	// 1 ≤ N ≤ 26
int tree[26][2];

void preorder(int cur) {
	if (cur == -19) return;			// '.' - 'A' = -19 
	cout << (char)(cur + 'A');      // 1. 루트
	preorder(tree[cur][0]);         // 2. 왼쪽
	preorder(tree[cur][1]);         // 3. 오른쪽
}

void inorder(int cur) {
	if (cur == -19) return;
	inorder(tree[cur][0]);          // 1. 왼쪽
	cout << (char)(cur + 'A');      // 2. 루트
	inorder(tree[cur][1]);          // 3. 오른쪽
}

void postorder(int cur) {
	if (cur == -19) return;
	postorder(tree[cur][0]);        // 1. 왼쪽
	postorder(tree[cur][1]);        // 2. 오른쪽
	cout << (char)(cur + 'A');      // 3. 루트
}

int main() {
	
	ios::sync_with_stdio(false);
	cin.tie(0);

	cin >> N;

	for(int i = 0; i < N; i++)	{
		char parent, left, right;
		cin >> parent >> left >> right;

		tree[parent - 'A'][0] = left - 'A';
		tree[parent - 'A'][1] = right - 'A';
	}

	preorder(0);
	cout << '\n';

	inorder(0);
	cout << '\n';

	postorder(0);
	cout << '\n';

	return 0;
}

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    
    static int tree[][];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        // A : 0 으로 해싱, [][0] : 왼쪽    [][1] : 오른쪽
        tree = new int[26][2];

        while(N-- > 0)
        {
            StringTokenizer st = new StringTokenizer(br.readLine());
            // 각 알파벳에서 A 의 아스키 값인 65를 빼주어 해싱해준다.  A : 0, B : 1, ...
            int parent = (int) st.nextToken().charAt(0) - 65;
            int left = (int) st.nextToken().charAt(0) - 65;
            int right = (int) st.nextToken().charAt(0) - 65;

            tree[parent][0] = left;
            tree[parent][1] = right;
        }

        preorder(0);
        System.out.println();

        inorder(0);
        System.out.println();

        postorder(0);
        System.out.println();
    }

    static void preorder(int c) {
        if(c == -19)    return;                 // '.' - 'A' = -19
        System.out.print((char) (c + 65));      // 1. 루트
        preorder(tree[c][0]);                   // 2. 왼쪽
        preorder(tree[c][1]);                   // 3. 오른쪽
    }

    static void inorder(int c)  {
        if(c == -19)    return;
        inorder(tree[c][0]);                    // 1. 왼쪽
        System.out.print((char) (c + 65));      // 2. 루트
        inorder(tree[c][1]);                    // 3. 오른쪽
    }

    static void postorder(int c)    {
        if(c == -19)    return;
        postorder(tree[c][0]);                  // 1. 왼쪽
        postorder(tree[c][1]);                  // 2. 오른쪽
        System.out.print((char) (c + 65));      // 3. 루트
    }
}